client/src/test/java/io/pravega/client/stream/impl/SegmentSelectorTest.java

```diff
@@ -217,4 +222,66 @@ public void testControllerNotReachable() {
         assertFutureThrows("Writer Future", writerFuture, t -> t instanceof ControllerFailureException);
     }
 
+    @Test
+    public void testSegmentRefreshOnSealed() {
+        final Segment segment0 = new Segment(scope, streamName, 0);
+        final Segment segment1 = new Segment(scope, streamName, computeSegmentId(1, 1));
+        final Segment segment2 = new Segment(scope, streamName, computeSegmentId(2, 1));
+        final CompletableFuture<Void> writerFuture = new CompletableFuture<>();
+        final PendingEvent pendingEvent = PendingEvent.withHeader("0", ByteBuffer.wrap("e".getBytes()), writerFuture);
+
+        TreeMap<Double, SegmentWithRange> segmentBeforeScale = new TreeMap<>();
+        addNewSegment(segmentBeforeScale, 0, 0.0, 1.0);
+        StreamSegments streamSegmentsBeforeScale = new StreamSegments(segmentBeforeScale, "");
+
+        Map<SegmentWithRange, List<Long>> newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(segment1, 0.0, 0.5), ImmutableList.of(segment0.getSegmentId()));
+        newRange.put(new SegmentWithRange(segment2, 0.5, 1.0), ImmutableList.of(segment0.getSegmentId()));
+        StreamSegmentsWithPredecessors segmentsWithPredecessors = new StreamSegmentsWithPredecessors(newRange, "");
+
+        // Setup Mock.
+        SegmentOutputStream s0Writer = Mockito.mock(SegmentOutputStream.class);
+        SegmentOutputStream s1Writer = Mockito.mock(SegmentOutputStream.class);
+        SegmentOutputStream s2Writer = Mockito.mock(SegmentOutputStream.class);
+        SegmentOutputStreamFactory factory = Mockito.mock(SegmentOutputStreamFactory.class);
+        Controller controller = Mockito.mock(Controller.class);
+
+        when(s0Writer.getUnackedEventsOnSeal()).thenReturn(ImmutableList.of(pendingEvent));
```

Comment:
Based on the observations of #3885 , should we be using `doAnswer` instead?

Reply:
The reason for the issue in #3885 is because are trying to stub operations on a `org.mockito.Mockito#spy(T)` of a real object. Hence doing a `when(spy.methodA())).thenReturn("foo");` would imply invoking methodA() while setting up the stubbed return value. 
In this case `s0Writer` is a mock object and hence we can use `when(...).thenReturn(...)` pattern.